### PR TITLE
Improve performance

### PR DESCRIPTION
**Pull request by jhogberg**
This PR implements some of the suggestions made in https://github.com/erlang/otp/pull/2143#issuecomment-469664033, almost doubling encode performance in [poison](https://github.com/devinus/poison)'s benchmark suite. Decoding is improved as well but not nearly as much as it was difficult to optimize without changing its semantics.

The changes made to trapping means that OTP `17.3` or higher is required, but it shouldn't be an issue given that OTP dropped support for that version 2 years ago.

[bench_before.txt](https://github.com/davisp/jiffy/files/3058617/bench_before.txt)
[bench_after.txt](https://github.com/davisp/jiffy/files/3058618/bench_after.txt)

### CODE REVIEW

**Comment by lpgauth** [comment]
@jhogberg Hi, for of all thanks for this patch! 

When you say `without changing its semantics`. What semantic changes would allow you to further optimize the decoding?

**Comment by lpgauth** [comment]
FYI, tried your branch on a production node and it segfaults. I'll try to run it inside GDB later today to find where it's happening.

**Comment by davisp** [comment]
Thanks for the PR! Totally forgot to make a note that I'd seen this but I am trying to find time to do a thorough review of everything. I skimmed it and everything seemed fine at first glance but I want to spend some more time considering various aspects before merging it all in one go.

**Comment by jhogberg** [comment]
> When you say without changing its semantics. What semantic changes would allow you to further optimize the decoding?

Removing `copy_strings` and making it the default, handling duplicate keys by letting the last one win, having slightly less pedantic UTF-8 validation, et cetera, but I don't think you could make these changes and still call it `jiffy`. Essentially I'd need to make more or less the same trade-offs that the BIFs in https://github.com/erlang/otp/pull/2143 did and at that point I might as well just port those to the NIF API.

> FYI, tried your branch on a production node and it segfaults. I'll try to run it inside GDB later today to find where it's happening.

Thanks for pointing this out, I should've run `valgrind` before posting this. It found an out-of-bounds read in the decoder that could cause segfaults on invalid input. I find it somewhat unlikely to have caused your most recent segfaults since it affects the current version as well, but you can try the fix I pushed a few minutes ago.

> Thanks for the PR! Totally forgot to make a note that I'd seen this but I am trying to find time to do a thorough review of everything. I skimmed it and everything seemed fine at first glance but I want to spend some more time considering various aspects before merging it all in one go.

No problem, take your time.

**Comment by jhogberg** [comment]
I've pushed another few commits that cleans things up and makes encoding long strings a bit faster (50% in `GovTrack` and `Issue 90`, the others are more or less unchanged).

I've started looking further into decoding, and I'm thinking about using erlang/otp#2143's method of keeping the terms we haven't finished decoding yet in a stack (if we're in the middle of an object for example). It ought to improve performance without changing any semantics, but I'm not sure if I can find the time to implement it any time soon. I might turn it into a separate PR later on.

**Comment by lpgauth** [comment]
Hi, looks like the encode returns a `iolist()` now?

```erlang
jiffy:encode({[{<<"hello">>, <<"world">>}]}).
[<<"{\"hello\":\"world\"}">>]
```

vs

```erlang
jiffy:encode({[{<<"hello">>, <<"world">>}]}).
<<"{\"hello\":\"world\"}">>
```

**Comment by jhogberg** [comment]
The spec says `iodata()` and it used to produce a `binary()` on some inputs and an `iolist()` on others, the following returns an `iolist()` on master as well:

```erlang
jiffy:encode({[{hello,1 bsl 63}]}).
[<<"{\"hello\":">>,<<"9223372036854775808">>,<<"}">>]
```

Ironically the last few commits broke this example, so I've pushed a fix for that.

**Comment by lpgauth** [comment]
Having trouble getting the debug symbols for Jiffy but here's the backtrace:

```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7f757b3e7700 (LWP 34288)]
0x00000000005f8718 in enif_is_binary (env=0x7f757b3e6e20, term=0) at beam/erl_nif.c:1091
1091	beam/erl_nif.c: No such file or directory.
(gdb) backtrace
#0  0x00000000005f8718 in enif_is_binary (env=0x7f757b3e6e20, term=0) at beam/erl_nif.c:1091
#1  0x00007f74b095a415 in encode_iter () from /usr/share/rtb-gateway/_build/default/lib/jiffy/priv/jiffy.so
#2  0x00007f74b0961ada in encode_init () from /usr/share/rtb-gateway/_build/default/lib/jiffy/priv/jiffy.so
#3  0x0000000000457dfa in process_main () at x86_64-pc-linux-gnu/opt/smp/beam_cold.h:119
#4  0x00000000004670dc in sched_thread_func () at beam/erl_process.c:8469
#5  0x00000000006a1247 in thr_wrapper (vtwd=0x7fff4c8a3d70) at pthread/ethread.c:118
#6  0x00007f76d6eb7494 in start_thread () from /lib64/libpthread.so.0
#7  0x00007f76d69f45dd in clone () from /lib64/libc.so.6
```

**Comment by davisp** [comment]
@lpgauth Is the JSON you're decoding consistently causing the issue and something you could share? That'd make things easier to debug locally if possible.

**Comment by lpgauth** [comment]
@davisp this happens in production... we call encode 50K/s so I can't really pin point which term is causing it to crash.

**Comment by jhogberg** [comment]
It should be possible to see which term it is in the core dump. The frame that’s in `process_main` (#3 in this case) has the original arguments in the `reg` array, and they can be printed with macros you can find in `erts/etc/unix/etp-commands`. e.g.

```
source /home/foobar/otp/erts/etc/unix/etp-commands
etp reg[0]
```

**Comment by lpgauth** [comment]
@jhogberg 
```
(gdb) frame 3
#3  0x0000000000457dfa in process_main () at x86_64-pc-linux-gnu/opt/smp/beam_cold.h:119
119	x86_64-pc-linux-gnu/opt/smp/beam_cold.h: No such file or directory.
(gdb) etp reg[0]
{[{id,#HeapBinary<0x13,0x33323536,0x34393235,0x393432>},{seatbid,[{[{bid,[{[{id,#SubBinary<0x38,0,0x10000:0x69b89ba2>},{impid,#HeapBinary<0x13,0x39323635,0x35383730,0x3363636>},{price,0.849000},{adid,...},{...}]}]},{seat,#HeapBinary<0x4,0x38363131>}]}]},{cur,#HeapBinary<0x3,0x445355>}]}.
```

**Comment by garazdawi** [comment]
Do:
> set $etp_max_depth = 2000
> set $etp_max_string_length = 10000

To get rid of the `...` in the etp printout. Also John is unavilable at the moment, so it may be a while until he gets back to this.

**Comment by jhogberg** [comment]
> @jhogberg
> 
> ```
> (gdb) frame 3
> #3  0x0000000000457dfa in process_main () at x86_64-pc-linux-gnu/opt/smp/beam_cold.h:119
> 119	x86_64-pc-linux-gnu/opt/smp/beam_cold.h: No such file or directory.
> (gdb) etp reg[0]
> {[{id,#HeapBinary<0x13,0x33323536,0x34393235,0x393432>},{seatbid,[{[{bid,[{[{id,#SubBinary<0x38,0,0x10000:0x69b89ba2>},{impid,#HeapBinary<0x13,0x39323635,0x35383730,0x3363636>},{price,0.849000},{adid,...},{...}]}]},{seat,#HeapBinary<0x4,0x38363131>}]}]},{cur,#HeapBinary<0x3,0x445355>}]}.
> ```

@lpgauth: Thank you, and sorry about the wait. The problem was that a `memcpy` disappeared while refactoring, it should be more stable now.

**Comment by lpgauth** [comment]
@jhogberg I'm on vacation for the next week, will test when I'm back. Thanks!

**Comment by jhogberg** [comment]
Cool, have a good one! :)

**Comment by lpgauth** [comment]
@jhogberg I confirm that the segfault is fixed. The optimizations make a noticeable change for encoding, unfortunately performance of decoding seems unchanged.

jiffy_decode

**Comment by jhogberg** [comment]
Great, I'll try to make another PR that aims to improve decode performance once this one's finished.

**Comment by lpgauth** [comment]
@jhogberg 👍 awesome, thanks

**Comment by lpgauth** [comment]
@davisp Hey, any plan on reviewing / merging this PR?

Thanks in advance :)

**Comment by davisp** [comment]
Yeah, sorry! Been busying with other things. Will try and spend some time today on reviewing it during my ops shift.

**Comment by davisp** [comment]
Have read through again much closer this time. I'm fairly happy with everything. I stumbled over the atom latin1/utf-8 comment for awhile before I realized you meant Latin1 is defined as identical to the Unicode code points.

There's a few style things I'm going to tweak for consistency and I'm also going to move the yielding logic so that I can if-def it out for older Erlang VMs.

Awesome work!

File: c_src/termstack.c [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    int arity;
+
+    if(enif_get_tuple(env, from, &arity, &elements)) {
+        stack->top = arity;
+
+        if(arity <= SMALL_TERMSTACK_SIZE) {
+            stack->elements = &stack->__default_elements[0];
+            stack->size = SMALL_TERMSTACK_SIZE;
+        } else {
+            stack->size = arity * 2;
```
Comment (davisp) [inline c_src/termstack.c]: What's the reasoning for double the stack here on restore? There's already a double each time we need more room. It seems bit odd to do this double on every iteration regardless of whether we need to or not.

Comment (davisp) [reply c_src/termstack.c]: Derp. Just realized its cause you truncate the stack on save so we'd immediately double on the next push.

**Comment by davisp** [comment]
Am nearly done reviewing and testing. I should be able to get this work merged tomorrow. The only thing that concerns me is the default output of a list change. I may go and apply the other PR for changing throw exceptions to error exceptions and call this 1.0.

Yet again, thanks for the awesome work on this @jhogberg!

**Comment by lpgauth** [comment]
`call this 1.0` 🎉 Thanks again @jhogberg & @davisp

**Comment by jhogberg** [comment]
> Am nearly done reviewing and testing. I should be able to get this work merged tomorrow. The only thing that concerns me is the default output of a list change. I may go and apply the other PR for changing throw exceptions to error exceptions and call this 1.0.
> 
> Yet again, thanks for the awesome work on this @jhogberg!

Sounds good, I'm glad I could help! :)

**Comment by davisp** [comment]
Merged by hand.

**Pull request closed without merging**